packages/compass-components/src/components/bson-document-editor.tsx

```diff
@@ -0,0 +1,180 @@
+    text: string;
+    value: BSONValue;
+  } | null>(null);
+  const [currentText, setCurrentText] = useState(text ?? '');
+
+  function onChangeTextWrapper(text: string, event?: any) {
+    setCurrentText(text);
+    onChangeText?.(text, event);
+    if (onChangeValue) {
+      let value;
+      let error: Error | null = null;
+      try {
+        const parsed = queryParser(currentText);
+        if (!parsed || typeof parsed !== 'object') {
+          // XXX(COMPASS-5689): We've hit the condition in
+          // https://github.com/mongodb-js/ejson-shell-parser/blob/c9c0145ababae52536ccd2244ac2ad01a4bbdef3/src/index.ts#L36
+          // in which instead of returning a parsed value or throwing an error,
+          // the query parser just returns an empty string when encountering
+          // input that can be parsed as JS but not as a valid query.
+          // Unfortunately, this also means that all context around what
+          // caused this error is unavailable here.
+          throw new Error('Field contained invalid input');
+        }
+
+        if (currentVariant === 'Shell') {
+          value = parsed;
+        } else {
+          value = EJSON.deserialize(parsed, { relaxed: false });
+        }
+      } catch (err) {
+        error = err as Error;
+      }
+      onChangeValue(value, error, currentVariant, text, event);
+    }
+  }
+
+  function onToggle() {
+    const newVariant = currentVariant === 'Shell' ? 'EJSON' : 'Shell';
+    setCurrentVariant(newVariant);
+    try {
+      let parsed = queryParser(currentText);
+      if (currentVariant === 'EJSON') {
+        parsed = EJSON.deserialize(parsed);
+      }
+
+      if (!parsed || typeof parsed !== 'object') {
+        throw new Error();
+      }
+      setLastAlternateVariantRepr({ text: currentText, value: parsed });
+
+      if (
+        EJSON.stringify(lastAlternateVariantRepr?.value, { relaxed: false }) ===
+          EJSON.stringify(parsed, { relaxed: false }) &&
+        lastAlternateVariantRepr?.text
+      ) {
+        setCurrentText(lastAlternateVariantRepr?.text);
+        return;
+      }
+
+      if (newVariant === 'EJSON') {
+        setCurrentText(queryParser.toJSString(EJSON.serialize(parsed)));
+      } else {
+        setCurrentText(queryParser.toJSString(parsed));
+      }
+    } catch {
+      /* do not replace text if not easily possible */
+    }
+  }
+
+  return (
+    <div style={{ position: 'relative' }}>
+      <div
+        ref={containerRef}
+        className={actionsGroupContainer}
+        style={{
+          display: isHovered ? 'flex' : 'none',
+        }}
+      >
+        <span
+          className={css({
+            flex: '1 0 auto',
+          })}
+        ></span>
+        <Button
+          size="xsmall"
+          title={
+            currentVariant === 'EJSON'
+              ? 'Switch to Shell BSON syntax'
+              : 'Switch to Extended JSON syntax'
+          }
+          aria-label={
```

Comment:
Maybe add a `title` in addition to the label so that you can also see the hint when you hover over the button

Reply:
Oh, sorry, don't know how I missed that it's literally one line above 🤦